Modules/atexitmodule.c

```diff
@@ -278,11 +278,20 @@ atexit_unregister(PyObject *module, PyObject *func)
             continue;
         }
 
+        /*
+         * Increment refcounts of func and cb->func because equality check may
+         * unregister one or both
+         */
+        PyObject *cb_func = Py_NewRef(cb->func);
+        PyObject *cur_func = Py_NewRef(func);
         int eq = PyObject_RichCompareBool(cb->func, func, Py_EQ);
+        Py_DECREF(cb_func);
+        Py_DECREF(cur_func);
         if (eq < 0) {
             return NULL;
         }
-        if (eq) {
+        // Equality comparison might have already deleted this callback
+        if (eq && state->callbacks[i] != NULL) {
```

Comment:
I think that it is better to check the refcount and not to rely on GC.

Reply:
I'm not sure what you mean. `state->callbacks[i]` is not a Python object, it's a pointer to

```C
typedef struct {
    PyObject *func;
    PyObject *args;
    PyObject *kwargs;
} atexit_py_callback;
```

So I don't think it would have a refcount. And `state->callbacks[i]` gets deleted and set to `NULL` deterministically in `atexit_delete_cb`, not by GC.